runtime/src/main/java/org/corfudb/runtime/Backup.java

```diff
@@ -0,0 +1,236 @@
+     *
+     * @throws IOException
+     */
+    private void backup() throws IOException {
+        for (UUID streamId : streamIDs) {
+            String fileName = backupTempDirPath + File.separator + streamId;
+            backupTable(fileName, streamId);
+        }
+        log.debug("successfully backed up {} tables to {} directory", streamIDs.size(), backupTempDirPath);
+    }
+
+    /**
+     * Back up a single table
+     *
+     * If the log is trimmed at timestamp, the backupTable will fail.
+     * If the table has no data to backupTable, it will create a file with empty contents.
+     *
+     * @param fileName   - the name of the backup file
+     * @param uuid       - the uuid of the table which is being backed up
+     * @throws IOException
+     */
+    private void backupTable(String fileName, UUID uuid) throws IOException {
+        try (FileOutputStream fileOutput = new FileOutputStream(fileName)) {
+            StreamOptions options = StreamOptions.builder()
+                    .ignoreTrimmed(false)
+                    .cacheEntries(false)
+                    .build();
+            Stream<OpaqueEntry> stream = (new OpaqueStream(runtime, runtime.getStreamsView().get(uuid, options))).streamUpTo(timestamp);
+
+            writeTableToFile(fileOutput, stream, uuid);
+        } catch (IOException e) {
+            log.error("failed to backup table {} to file {}", uuid, fileName);
+            throw e;
+        }
+
+        log.debug("{} table is backed up and stored to temp file {}", uuid, fileName);
+    }
+
+    private void writeTableToFile(FileOutputStream fileOutput, Stream<OpaqueEntry> stream, UUID uuid) throws IOException {
+        Iterator<OpaqueEntry> iterator = stream.iterator();
+        while (iterator.hasNext()) {
+            OpaqueEntry lastEntry = iterator.next();
+            List<SMREntry> smrEntries = lastEntry.getEntries().get(uuid);
+            if (smrEntries != null) {
+                Map<UUID, List<SMREntry>> map = new HashMap<>();
+                map.put(uuid, smrEntries);
+                OpaqueEntry newOpaqueEntry = new OpaqueEntry(lastEntry.getVersion(), map);
+                OpaqueEntry.write(fileOutput, newOpaqueEntry);
+            }
+        }
+        fileOutput.flush();
+    }
+
+    /**
+     * All generated files under tmp directory will be composed into one tar file
+     *
+     * @throws IOException
+     */
+    private void generateTarFile() throws IOException {
+        File folder = new File(backupTempDirPath);
+        File[] srcFiles = folder.listFiles();
+        if (srcFiles == null) {
+            log.debug("no backup file found under directory {}", backupTempDirPath);
+            return;
+        }
+
+        try (FileOutputStream fileOutput = new FileOutputStream(filePath);
+             TarArchiveOutputStream tarOutput = new TarArchiveOutputStream(fileOutput)) {
+            for (File srcFile : srcFiles) {
+                addToTarFile(srcFile, tarOutput);
```

Comment:
do we want to preserve the file or delete original after in the TAR? or is this done in the cleanup?

Reply:
Those files are under `backupTempDirPath` and are deleted in `cleanup()` at the end of the whole process.